pyvrp/IteratedLocalSearch.py

```diff
@@ -1 +1,256 @@
-# TODO
+from __future__ import annotations
+
+import time
+from collections import deque
+from dataclasses import dataclass
+from statistics import fmean
+from typing import TYPE_CHECKING
+
+from pyvrp.ProgressPrinter import ProgressPrinter
+from pyvrp.Result import Result
+from pyvrp.Statistics import Statistics
+
+if TYPE_CHECKING:
+    from pyvrp.PenaltyManager import PenaltyManager
+    from pyvrp._pyvrp import (
+        CostEvaluator,
+        ProblemData,
+        RandomNumberGenerator,
+        Solution,
+    )
+    from pyvrp.search.SearchMethod import SearchMethod
+    from pyvrp.stop.StoppingCriterion import StoppingCriterion
+
+
+@dataclass
+class IteratedLocalSearchParams:
+    """
+    Parameters for the iterated local search algorithm.
+
+    Parameters
+    ----------
+    num_iters_no_improvement
+        Number of iterations without any improvement needed before a restart
+        occurs.
+    initial_accept_weight
+        Initial weight parameter used to determine the threshold value in the
+        acceptance criterion. Larger values result in more accepted candidate
+        solutions. Must be in [0, 1].
+    history_length
+        The number of recent candidate solutions to consider when computing the
+        threshold value in the acceptance criterion. Must be positive.
+    """
+
+    num_iters_no_improvement: int = 20_000
+    initial_accept_weight: float = 1
+    history_length: int = 500
+
+    def __post_init__(self):
+        if self.num_iters_no_improvement < 0:
+            raise ValueError("num_iters_no_improvement < 0 not understood.")
+
+        if not (0 <= self.initial_accept_weight <= 1):
+            raise ValueError("initial_accept_weight must be in [0, 1].")
+
+        if self.history_length <= 0:
+            raise ValueError("history_length must be positive.")
+
+
+class IteratedLocalSearch:
+    """
+    Creates an IteratedLocalSearch instance.
+
+    Parameters
+    ----------
+    data
+        The problem data instance.
+    penalty_manager
+        Penalty manager to use.
+    rng
+        Random number generator.
+    search_method
+        Search method to use.
+    initial_solution
+        Initial solution to start the search with.
+    params
+        Iterated local search parameters to use. If not provided, a default
+        will be used.
+    """
+
+    def __init__(
+        self,
+        data: ProblemData,
+        penalty_manager: PenaltyManager,
+        rng: RandomNumberGenerator,
+        search_method: SearchMethod,
+        initial_solution: Solution,
+        params: IteratedLocalSearchParams = IteratedLocalSearchParams(),
+    ):
+        self._data = data
+        self._pm = penalty_manager
+        self._rng = rng
+        self._search = search_method
+        self._init = initial_solution
+        self._params = params
+
+        self._history: deque[int] = deque(maxlen=self._params.history_length)
```

Comment:
Reverted to a double-ended queue for the time being, as it makes things much cleaner. In future PRs, we should probably write our own custom `SolutionHistory` class in C++.